server/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1741,4 +1605,98 @@ public final ResponseHandlers getResponseHandlers() {
         return requestHandlers.get(action);
     }
+
+    private final class ChannelsConnectedListener implements ActionListener<Void> {
+
+        private static final int FAILED = -1;
+
+        private final DiscoveryNode node;
+        private final ConnectionProfile connectionProfile;
+        private final List<TcpChannel> channels;
+        private final ActionListener<NodeChannels> listener;
+        private final AtomicInteger pendingConnections;
+
+        private ChannelsConnectedListener(DiscoveryNode node, ConnectionProfile connectionProfile, List<TcpChannel> channels,
+                                          ActionListener<NodeChannels> listener) {
+            this.node = node;
+            this.connectionProfile = connectionProfile;
+            this.channels = channels;
+            this.listener = listener;
+            this.pendingConnections = new AtomicInteger(channels.size());
+        }
+
+        @Override
+        public void onResponse(Void v) {
+            assert pendingConnections.get() != 0 : "Should not call onResponse when the pending connections is 0.";
+            // Returns true if all connections have completed successfully
+            if (setConnectSuccess()) {
+                final TcpChannel handshakeChannel = channels.get(0);
+                try {
+                    executeHandshake(node, handshakeChannel, connectionProfile.getHandshakeTimeout(), new ActionListener<Version>() {
+                        @Override
+                        public void onResponse(Version version) {
+                            NodeChannels nodeChannels = new NodeChannels(node, channels, connectionProfile, version);
+                            nodeChannels.channels.forEach(ch -> ch.addCloseListener(ActionListener.wrap(nodeChannels::close)));
+                            listener.onResponse(nodeChannels);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            CloseableChannel.closeChannels(channels, false);
+
+                            if (e instanceof ConnectTransportException) {
+                                listener.onFailure(e);
+                            } else {
+                                listener.onFailure(new ConnectTransportException(node, "general node connection failure", e));
+                            }
+                        }
+                    });
+                } catch (Exception ex) {
+                    CloseableChannel.closeChannels(channels, false);
+                    listener.onFailure(ex);
+                }
+            }
+        }
+
+        @Override
+        public void onFailure(Exception ex) {
+            assert pendingConnections.get() != 0 : "Should not receive non-timeout connection exception if no connections pending.";
+            if (setFailed()) {
+                CloseableChannel.closeChannels(channels, false);
```

Comment:
`CloseableChannel.closeChannel` can throw so should we call the listener first or use try/finally?